File: py/src/playwright_adventures/journeys/journey_types.py
from typing import Literal, TypedDict


class TextPattern(TypedDict):
    pattern: str
    ignoreCase: bool


class RoleSelector(TypedDict):
    first: bool
    by: Literal["role"]
    role: Literal["button", "heading", "link"]
    name: TextPattern | None
    level: int | None


class LabelSelector(TypedDict):
    first: bool
    by: Literal["label"]
    name: TextPattern


class TestIdSelector(TypedDict):
    first: bool
    by: Literal["testId"]
    value: str


SelectorSpec = RoleSelector | LabelSelector | TestIdSelector
FixtureValue = Literal["user.email", "user.password", "user.displayName"]


class NavigateStep(TypedDict):
    action: Literal["navigate"]
    path: str


class ClickStep(TypedDict):
    action: Literal["click"]
    selector: SelectorSpec


class FillStep(TypedDict):
    action: Literal["fill"]
    selector: SelectorSpec
    value: FixtureValue


class AssertVisibleStep(TypedDict):
    action: Literal["assert-visible"]
    selector: SelectorSpec


JourneyStep = NavigateStep | ClickStep | FillStep | AssertVisibleStep


class JourneySpec(TypedDict):
    id: str
    name: str
    description: str
    successMessage: str
    extends: str | None
    preconditions: list[str]
    steps: list[JourneyStep]

File: js/src/lib/journeys/journeyTypes.ts
export interface TextPattern {
  pattern: string;
  ignoreCase: boolean;
}

interface SelectorBase {
  first: boolean;
}

export interface RoleSelector extends SelectorBase {
  by: 'role';
  role: 'button' | 'heading' | 'link';
  name: TextPattern | null;
  level: number | null;
}

export interface LabelSelector extends SelectorBase {
  by: 'label';
  name: TextPattern;
}

export interface TestIdSelector extends SelectorBase {
  by: 'testId';
  value: string;
}

export type SelectorSpec = RoleSelector | LabelSelector | TestIdSelector;
export type FixtureValue = 'user.email' | 'user.password' | 'user.displayName';

export interface NavigateStep {
  action: 'navigate';
  path: string;
}

export interface ClickStep {
  action: 'click';
  selector: SelectorSpec;
}

export interface FillStep {
  action: 'fill';
  selector: SelectorSpec;
  value: FixtureValue;
}

export interface AssertVisibleStep {
  action: 'assert-visible';
  selector: SelectorSpec;
}

export type JourneyStep = NavigateStep | ClickStep | FillStep | AssertVisibleStep;

export interface JourneySpec {
  id: string;
  name: string;
  description: string;
  successMessage: string;
  extends: string | null;
  preconditions: readonly string[];
  steps: readonly JourneyStep[];
}
